### DISCUSSION TITLE
Option to Automatically Rollback to a Snapshot

### DISCUSSION THREAD

**Discussion by michacassola** [zbm-dev/zfsbootmenu#679] [Feature Requests]
I would like it if ZBM could be configured to prefer a certain pool's snapshot and boot into that and if that fails boots into another somehow marked pool snapshot.

This way the running system could become the fallback snapshot that is done before an upgrade.

If this is out of scope for ZBM, how would I best go about making something like that, if it is at all possible?

**Comment by ahesford** [comment]
This is definitely out of scope, and is going to involve a lot of handshaking between your regular boot environment and ZFSBootMenu. You can do the ZBM side from an early-setup hook or add a custom startup script in /usr/share/zfsbootmenu/init.d. Roughly:
1. You'll probably need to import the pool with write privileges in ZBM. We import readonly by default.
2. Check some ZFS property, like `your.custom.domain:boot-attempt`, on your boot environment.
    - If this property is already set, you can assume a boot failed, so you'll need to figure out what prior snapshot you want to clone (you can't really boot a snapshot, but you can clone it to a regular filesystem and boot that), do the clone, and set the default boot for this iteration. (If you're doing this in early setup, just set `bootfs` on your pool and we'll pick it up. If you're doing this with a custom `init.d` script after ZBM already imported your pools, you can overwrite `${BASE}/bootfs` to point to the new clone. You'll *really* need to understand ZBM logic to hijack this process.) You probably want to look at another property, like `your.custom.domain:last-boot`, to see if you know what was successfully booted last, and use that to inform things. (See the next step.)
    - If the property is not set, you must have had a successful boot, so record the current default in `your.custom.domain:boot-attempt` and let ZBM do its thing.
3. Make your boot environment run a script that clears `your.custom.domain:boot-attempt`, and maybe sets something like `your.custom.domain:last-boot` to indicate what actually did boot.

This will probably get pretty hacky and highly customized, and I don't know that it's advisable to attempt to automate recovery of this nature. It's probably better to just build a ZBM image with network connectivity and SSH access, so that you can log in even from afar and just pick what you want to do when you need to do it.
